extensions/pkg/terraformer/terraformer.go

```diff
@@ -141,7 +141,16 @@ func (t *terraformer) Destroy(ctx context.Context) error {
 	if err := t.execute(ctx, "destroy"); err != nil {
 		return err
 	}
-	return t.CleanupConfiguration(ctx)
+
+	if err := t.CleanupConfiguration(ctx); err != nil {
+		return err
+	}
+
+	if !t.resourcesAreFinalized(ctx) {
+		return fmt.Errorf("terraformer finalizers are still present, needs to retry")
```

Comment:
Not sure, whether a retry will always help, as terraformer refuses to execute when one of the resources is missing, e.g. the `tf-config`.
Though, in some cases it might help, so no objection against proceeding like this.

Reply:
Valid point. It can be in deadlock here. Cannot delete `Infrastructure` because of the still existing configmap/secret and cannot delete the configmaps/secrets because not all of them are present at the same time.

I am starting to think it was a mistake to let the terraformer manage the finalizers on these resources, mainly because it is not an active controller watching the resources, but just a one-time job per Infrastructure reconciliation. 

What do you think about managing the finalizer from within the Infrastructure controller?

Reply:
Oh yeah, true. Didn't notice that it will even be a deadlock. That's not good.



> I am starting to think it was a mistake to let the terraformer manage the finalizers on these resources, mainly because it is not an active controller watching the resources, but just a one-time job per Infrastructure reconciliation.
> 
> What do you think about managing the finalizer from within the Infrastructure controller?

Hmm, why do you think so? Terraformer is actually creating external infrastructure on behalf of the k8s resources, so it would be the correct component to also protect the resources with finalizers, don't you think?

Reply:
> Hmm, why do you think so? Terraformer is actually creating external infrastructure on behalf of the k8s resources, so it would be the correct component to also protect the resources with finalizers, don't you think?

I completely agree, and this was the reason why I voted to implement it this way back then.